### PR TITLE
add rerank, update demo and tests

### PR DESCRIPTION
**Pull request by sahel-sh**
# Pull Request Checklist

## Reference Issue

Please provide the reference to issue this PR is addressing (# followed by the issue number). If there is no associated issue, write "N/A".

ref: 

## Checklist Items

Before submitting your pull request, please review these items:

- [ ] Have you followed the [contributing guidelines](CONTRIBUTING.md)?
- [x] Have you verified that there are no existing [Pull Requests](https://github.com/castorini/rank_llm/pulls) for the same update/change?
- [ ] Have you updated any relevant documentation or added new tests where needed?

# PR Type #110 

What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Documentation content changes
- [ ] Other... 
    - Description:

### CODE REVIEW

File: src/rank_llm/rerank/rank_listwise_os_llm.py [hunk 1]
```diff
@@ -57,9 +57,9 @@ def __init__(
         self._device = device
         if self._device == "cuda":
             assert torch.cuda.is_available()
-        if prompt_mode != PromptMode.RANK_GPT:
+        if prompt_mode is not PromptMode.RANK_GPT:
```
Comment (ronakice) [inline src/rank_llm/rerank/rank_listwise_os_llm.py:60]: Is this preferred by standards here?

File: test/evaluation/test_trec_eval.py [hunk 2]
```diff
@@ -2,22 +2,22 @@
...
-                hits=[
-                    {"qid": "q1", "docid": "D1", "rank": 1, "score": 0.9},
-                    {"qid": "q1", "docid": "D2", "rank": 2, "score": 0.8},
+                candidates=[
+                    {"qid": "q1", "docid": "D1", "score": 0.9},
+                    {"qid": "q1", "docid": "D2", "score": 0.8},
                 ],
             ),
             Result(
                 query="Query2",
```
Comment (ronakice) [inline test/evaluation/test_trec_eval.py:19]: Isn't `query` different too?

**Review by ronakice — APPROVED** [review]

File: src/rank_llm/rerank/rank_listwise_os_llm.py (hunk 1 above)
Comment (sahel-sh) [reply src/rank_llm/rerank/rank_listwise_os_llm.py]: reverted it to the preview state

File: test/evaluation/test_trec_eval.py (hunk 2 above)
Comment (sahel-sh) [reply test/evaluation/test_trec_eval.py]: fixed

**Pull request merged**